### ISSUE TITLE
More rule of thumb info on hyperparameters

### ISSUE THREAD

**Issue by vqt123** [Unity-Technologies/ml-agents#1642]
Lots of good info on hyperparameters here:
https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Training-PPO.md

More info though that I have found useful that maybe useful for others starting off

**Batchsize / Buffersize:** The bigger, the more steps are required to train
**Hidden Units:** The bigger, the longer each step takes to process
**Num Layers:** The bigger, the longer each step takes to process
**Learning rate:** the bigger the faster the training. The slope is dependant on max steps, you need to adjust this if you adjust your max steps
**Gamma:** The bigger, the faster the training. Can you elaborate more on the wierd range.. why .9 to .99? Is it safe to set it to .5?
**Lamda:** Same as Gamma? What happens if we set it to .5? what happens if we set it to 2.0? The range for this value is also hard to understand.

**Comment by MarcoMeter** [comment]
`Batchsize / Buffersize: The bigger, the more steps are required to train`

Depending on the utilized environment, larger batches may shrink down the number of  training steps required to let the behavior converge to a minima. A large buffer could contain old experience tuples, which are not useful anymore for updating the agent's policy, thus more training steps may be required.

`Hidden Units: The bigger, the longer each step takes to process
Num Layers: The bigger, the longer each step takes to process`

These affect the size of the neural network. The more values (i.e. weights) are to be updated, the more time it will take to train for each step.

`Learning rate: the bigger the faster the training. The slope is dependant on max steps, you need to adjust this if you adjust your max steps`

The learning rate determines the strength of the updates of the aforementioned weights. A larger learning rate does stronger updates, but this might cause the network to always overshoot a desired minima. That's why the learning rate is reduced over time so that smaller steps can be done to find the minima.

This learning rate is setup by mlagents to linearly decay based on the number of trained steps. The learning rate does not have to be changed, because of choosing a different value for the max steps.

`Gamma: The bigger, the faster the training. Can you elaborate more on the wierd range.. why .9 to .99? Is it safe to set it to .5?`

The discount factor is specified by gamma. It rather alters the agent's behavior instead of impacting the training progression (time-wise). If you choose 0.5 for gamma, then your agent would choose more immediate rewards instead of planning to get much more rewards in the long run, if a bigger gamma was chosen.



To understand how neural networks learn, you could watch the small series by 3Blue1Brown:
https://youtu.be/aircAruvnKk

**Comment by MarcoMeter** [comment]
Decision Frequency:
The DecisionFrequency determines when a new decision is to be made by the agent. If it is set to 1, the agent will make a new decision every frame. If it is set to 3, the agent make a decision, repeats it for the next (I think) two frames and then makes a new decision on the third step.

The sequence length and memory size correspond to the recurrent neural network part of the to be trained model (as long as if you set recurrent to true in the hyperparameters). I didn't work with RNN yet, so I just refer to the docs: https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Training-PPO.md

**Comment by mattinjersey** [comment]
I also observe that when watching the Unity screen, it updates more frequently when the Decision Frequency is higher. Why does that occur.

**Comment by MarcoMeter** [comment]
The environment build with Unity gathers a trajectory of data (I think as large as the time horizon is) and then sends it to the training algorithm in Python. So the simulation runs for a bit to gather some data.

If data was sent step by step, the python interface would be a huge performance bottleneck.

**Comment by mattinjersey** [comment]
ok. will time horizon setting affect the results.

**Comment by mattinjersey** [comment]
Also going back to Decision Frequency, shall we assume 60 fps?

**Comment by vincentpierre** [comment]
Thank you for the discussion. We are closing this issue due to inactivity.

**Comment by MarcoMeter** [comment]
One more question on this context. How do the mini batch updates fit to your hyerparameter naming conventions? Like what is the actual batch size and the mini batch size in ml-agents?
